browser/component_updater/brave_component_contents_verifier.cc

```diff
@@ -0,0 +1,171 @@
+}
+
+// Proxies the file access and checks the verified_contents.json.
+class ComponentContentsAccessorImpl
+    : public brave_component_updater::ComponentContentsAccessor {
+ public:
+  explicit ComponentContentsAccessorImpl(const base::FilePath& component_root)
+      : brave_component_updater::ComponentContentsAccessor(component_root) {
+    if (base::PathExists(
+            component_root.AppendASCII(kBraveVerifiedContentsPath))) {
+      verified_contents_ = extensions::VerifiedContents::CreateFromFile(
+          kComponentContentsVerifierPublicKey,
+          component_root.AppendASCII(kBraveVerifiedContentsPath));
+    } else {
+      verified_contents_ = extensions::VerifiedContents::CreateFromFile(
+          kComponentContentsVerifierPublicKey,
+          component_root.AppendASCII(kVerifiedContentsPath));
+    }
+    if (verified_contents_ &&
+        verified_contents_->block_size() % crypto::kSHA256Length != 0) {
+      // Unsupported block size.
+      verified_contents_.reset();
+    }
+  }
+
+  bool IsComponentSignatureValid() const override {
+    return !!verified_contents_.get();
+  }
+
+  bool VerifyContents(const base::FilePath& relative_path,
+                      base::span<const uint8_t> contents) override {
+    if (!IsComponentSignatureValid()) {
+      return false;
+    }
+
+    if (!verified_contents_->HasTreeHashRoot(relative_path)) {
+      return false;
+    }
+
+    const auto root_hash =
+        GetRootHasheForContent(contents, verified_contents_->block_size());
+    return verified_contents_->TreeHashRootEquals(relative_path, root_hash);
+  }
+
+ private:
+  ~ComponentContentsAccessorImpl() override = default;
+
+  std::unique_ptr<extensions::VerifiedContents> verified_contents_;
+};
+
+bool ShouldBypassSignature() {
+  static const bool kBypass =
+      !base::FeatureList::IsEnabled(
+          component_updater::kComponentContentsVerifier) ||
+      base::CommandLine::ForCurrentProcess()->HasSwitch(
+          component_updater::kBypassComponentContentsVerifier);
+  return kBypass;
+}
+
+}  // namespace
+
+#endif  // BUILDFLAG(ENABLE_EXTENSIONS)
+
+namespace component_updater {
+
+BASE_FEATURE(kComponentContentsVerifier,
+             "ComponentContentsVerifier",
+             base::FEATURE_DISABLED_BY_DEFAULT);
+
+scoped_refptr<brave_component_updater::ComponentContentsAccessor>
+CreateComponentContentsAccessor(bool with_verifier,
+                                const base::FilePath& component_root) {
+#if BUILDFLAG(ENABLE_EXTENSIONS)
+  static const bool kBypassSignature = ShouldBypassSignature();
+  if (with_verifier && !kBypassSignature) {
+    return base::MakeRefCounted<ComponentContentsAccessorImpl>(component_root);
+  }
+#endif
+  // if there is no extensions enabled then we expect that on these platforms
+  // the component files are protected by the OS.
+  return base::MakeRefCounted<
+      brave_component_updater::ComponentContentsAccessor>(component_root);
```

Comment:
I'd prefer to save the `ComponentContentsAccessor` name for the base class, because it exposed to the consumers and it's strange to get the secure behaviour within the SimpleCA class.

Reply:
It's no longer the base class